indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/Asn1Decoding.kt

```diff
@@ -10,99 +10,171 @@ import at.asitplus.signum.indispensable.asn1.BERTags.T61_STRING
-fun Asn1Element.Companion.parse(source: ByteArray): Asn1Element = parse(source.iterator())
+fun Asn1Element.Companion.parse(source: ByteArray): Asn1Element =
+    source.wrapInUnsafeSource().readFullyToAsn1Elements().first.let {
+        if (it.size != 1)
+            throw Asn1StructuralException("Trailing bytes found after the first ASN.1 element")
+        it.first()
+    }
 
 /**
  * Tries to parse the [input] into a list of [Asn1Element]s. Consumes all Bytes and throws if an invalid ASN.1 Structure is found at any point.
  * @return the parsed elements
  *
  * @throws Asn1Exception on invalid input or if more than a single root structure was contained in the [input]
+ *
  */
+@Deprecated(
+    "Use a ByteArray or (even better) a kotlinx.io Source as input when possible. This method copies all bytes from the input twice and is inefficient.",
+    ReplaceWith("source.readFullyToAsn1Elements()"),
+    DeprecationLevel.WARNING
+)
 @Throws(Asn1Exception::class)
-fun Asn1Element.Companion.parseAll(input: ByteIterator): List<Asn1Element> = input.doParseAll()
+fun Asn1Element.Companion.parseAll(input: ByteIterator): List<Asn1Element> =
+    mutableListOf<Byte>().also { while (input.hasNext()) it.add(input.nextByte()) }.toByteArray().wrapInUnsafeSource()
+        .readFullyToAsn1Elements().first
 
 /**
  * Convenience wrapper around [parseAll], taking a [ByteArray] as [source]
  * @see parse
  */
 @Throws(Asn1Exception::class)
-fun Asn1Element.Companion.parseAll(source: ByteArray): List<Asn1Element> = parseAll(source.iterator())
+fun Asn1Element.Companion.parseAll(source: ByteArray): List<Asn1Element> =
+    source.wrapInUnsafeSource().readFullyToAsn1Elements().first
 
+/**
+ * Convenience wrapper around [parseFirst], taking a [ByteArray] as [source].
+ * @return a pair of the first parsed [Asn1Element] mapped to the remaining bytes
+ * @see readAsn1Element
+ */
+@Throws(Asn1Exception::class)
+fun Asn1Element.Companion.parseFirst(source: ByteArray): Pair<Asn1Element, ByteArray> =
+    source.wrapInUnsafeSource().readAsn1Element()
+        .let { Pair(it.first, source.copyOfRange(it.second.toInt(), source.size)) }
+
+private fun Source.doParseExactly(nBytes: Long): List<Asn1Element> = mutableListOf<Asn1Element>().also { list ->
+    var nBytesRead: Long = 0
+    while (nBytesRead < nBytes) {
+        val peekTagAndLen = peekTagAndLen()
+        val numberOfNextBytesRead = peekTagAndLen.second + peekTagAndLen.first.length
+        if (nBytesRead + numberOfNextBytesRead > nBytes) break
+        skip(peekTagAndLen.second.toLong()) // we only peeked before, so now we need to skip,
+        //                                     since we want to recycle the result below
+        val (elem, read) = readAsn1Element(peekTagAndLen.first, peekTagAndLen.second)
+        list.add(elem)
+        nBytesRead += read
+    }
+    require(nBytesRead == nBytes) { (if (nBytesRead < nBytes) "Less " else "More ") + " bytes read ($nBytesRead) than allowed ($nBytes)" }
```

Comment:
also says "fewer" now

Reply:
should maybe just say `indicated length ($nBytes) does not correspond to an asn.1 element boundary`, since i think that's what it's functionally used for?